[SPARK-24127][SS] Continuous text socket source

arunmahadevan opened the pull request:
## What changes were proposed in this pull request?

Support for text socket stream in spark structured streaming "continuous" mode. This is roughly based on the idea of ContinuousMemoryStream where the executor queries the data from driver over an RPC endpoint.

This makes it possible to create Structured streaming continuous pipeline to ingest data via "nc" and run examples.

## How was this patch tested?

Unit test and ran spark examples in structured streaming continuous mode.

Please review http://spark.apache.org/contributing.html before opening a pull request.

arunmahadevan committed:
SPARK-24127: Continuous text socket source

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

arunmahadevan committed:
SPARK-24127: Added unit test

arunmahadevan commented:
ping @jerryshao @tdas @jose-torres @HeartSaVioR for inputs.

HyukjinKwon commented:
ok to test

HyukjinKwon commented:
add to whitelist

SparkQA commented:
**[Test build #90349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90349/testReport)** for PR 21199 at commit [`f010943`](https://github.com/apache/spark/commit/f010943699b184cc9572bda8651cb40d6231bfa3).

SparkQA commented:
**[Test build #90349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90349/testReport)** for PR 21199 at commit [`f010943`](https://github.com/apache/spark/commit/f010943699b184cc9572bda8651cb40d6231bfa3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90349/
Test FAILed.

jose-torres commented:
I won't be able to look at this in detail until next week.

In general, I think this is a great source to have available. I wonder if it'd be worthwhile to try and abstract the record forwarding RPCs from here and ContinuousMemoryStream together.

xuanyuanking commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala:
```diff
@@ -0,0 +1,304 @@
+          context.reply(
+            record.map(r => if (includeTimestamp) Row(r) else Row(r._1)
+            )
+          )
```
```
context.reply(record.map(r => if (includeTimestamp) Row(r) else Row(r._1)))
```
Just in one line is ok I think.

xuanyuanking commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala:
```diff
@@ -0,0 +1,304 @@
+      case (offset, partition) =>
+        buckets(partition).trimStart(offset - startOffset.offsets(partition))
+    }
+    startOffset = endOffset
```
If I understand right, this commit will never enter in your added test case.

arunmahadevan committed:
SPARK-24127: More unit tests

jerryshao commented:
I was thinking if it is too overkill to receive data in the driver side and publish them to the executors via RPC? This might give user a wrong impression that data should be received in the driver side and published to the executors again.

Just my two cents.

jose-torres commented:
I think that's unavoidable if we want to have a socket source. The microbatch socket source has the same thing going on. I'd expect most people looking into implementation details of data sources will understand that they ought to read from executors in general.

arunmahadevan committed:
SPARK-24127: Refactor the RPC endpoint

SparkQA commented:
**[Test build #90603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90603/testReport)** for PR 21199 at commit [`b3a42f0`](https://github.com/apache/spark/commit/b3a42f08cba85b9bec11aaa3f75de298aa869204).

arunmahadevan committed:
SPARK-24127: rebase and refactor to handle the DataReader* renames

SparkQA commented:
**[Test build #90606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90606/testReport)** for PR 21199 at commit [`b962c3d`](https://github.com/apache/spark/commit/b962c3dbd1715b2d4fa03e65731e36697cf37ff1).

SparkQA commented:
**[Test build #90606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90606/testReport)** for PR 21199 at commit [`b962c3d`](https://github.com/apache/spark/commit/b962c3dbd1715b2d4fa03e65731e36697cf37ff1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90606/
Test PASSed.

SparkQA commented:
**[Test build #90603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90603/testReport)** for PR 21199 at commit [`b3a42f0`](https://github.com/apache/spark/commit/b3a42f08cba85b9bec11aaa3f75de298aa869204).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class ContinuousRecordPartitionOffset(partitionId: Int, offset: Int) extends PartitionOffset`
  * `case class GetRecord(offset: ContinuousRecordPartitionOffset)`
  * `class ContinuousRecordEndpoint(buckets: Seq[Seq[Any]], lock: Object)`

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90603/
Test PASSed.

arunmahadevan committed:
SPARK-24127: Rename to match ReaderFactory -> InputPartition

SparkQA commented:
**[Test build #90744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90744/testReport)** for PR 21199 at commit [`242bcdb`](https://github.com/apache/spark/commit/242bcdbacad2b3133447ba7e7cef100567ed3fd0).

arunmahadevan commented:
A gentle ping for review @jose-torres , @jerryshao , @xuanyuanking

jose-torres commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ContinuousRecordEndpoint.scala:
```diff
@@ -0,0 +1,61 @@
+ * A RPC end point for continuous readers to poll for
+ * records from the driver.
+ *
+ * @param buckets the data buckets
```
nit: probably better to elaborate on what these are for, Seq[Seq[Any]] and Object aren't very informative types

arunmahadevan replied:
Added more comments to clarify.

jose-torres commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala:
```diff
@@ -256,7 +258,101 @@ class TextSocketStreamSuite extends StreamTest with SharedSQLContext with Before
+    val data = scala.collection.mutable.ListBuffer[Int]()
+    val offsets = scala.collection.mutable.ListBuffer[Int]()
+    import org.scalatest.time.SpanSugar._
+    failAfter(5 seconds) {
```
Can't we use testStream for these tests?

arunmahadevan replied:
Probably we could use, but the `addSocketData` does not work for continuous source and thought the reader offsets could be validated better this way. (followed the approach in RateStreamProviderSuite)

SparkQA commented:
**[Test build #90744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90744/testReport)** for PR 21199 at commit [`242bcdb`](https://github.com/apache/spark/commit/242bcdbacad2b3133447ba7e7cef100567ed3fd0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class TextSocketContinuousInputPartition(`
  * `class TextSocketContinuousInputPartitionReader(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90744/
Test FAILed.

arunmahadevan committed:
SPARK-24127: Added more code comments

SparkQA commented:
**[Test build #90755 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90755/testReport)** for PR 21199 at commit [`68c5eed`](https://github.com/apache/spark/commit/68c5eed7c6d32db8a7a469a1970e6fa731a91c70).

SparkQA commented:
**[Test build #90755 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90755/testReport)** for PR 21199 at commit [`68c5eed`](https://github.com/apache/spark/commit/68c5eed7c6d32db8a7a469a1970e6fa731a91c70).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90755/
Test PASSed.

arunmahadevan commented:
ping @tdas @jose-torres

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #91615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91615/testReport)** for PR 21199 at commit [`68c5eed`](https://github.com/apache/spark/commit/68c5eed7c6d32db8a7a469a1970e6fa731a91c70).

SparkQA commented:
**[Test build #91615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91615/testReport)** for PR 21199 at commit [`68c5eed`](https://github.com/apache/spark/commit/68c5eed7c6d32db8a7a469a1970e6fa731a91c70).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91615/
Test PASSed.

HeartSaVioR commented:
@arunmahadevan 
Sorry I forgot to review this so far. Could you fix merge conflicts? I'd pull the code to the local and review since the code diff is not small.

SparkQA commented:
**[Test build #93797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93797/testReport)** for PR 21199 at commit [`76512d8`](https://github.com/apache/spark/commit/76512d8ad86ea9e7301db0a9ac9fce12a66d3f40).

SparkQA commented:
**[Test build #93797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93797/testReport)** for PR 21199 at commit [`76512d8`](https://github.com/apache/spark/commit/76512d8ad86ea9e7301db0a9ac9fce12a66d3f40).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93797/
Test FAILed.

arunmahadevan committed:
SPARK-24127: rebase and fix issues

arunmahadevan commented:
@HeartSaVioR , rebased with master. 

ping @jose-torres @tdas @zsxwing for review.

SparkQA commented:
**[Test build #93801 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93801/testReport)** for PR 21199 at commit [`a069d01`](https://github.com/apache/spark/commit/a069d01a485b6066650b75e0cb9bfaf63710a9e2).

SparkQA commented:
**[Test build #93801 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93801/testReport)** for PR 21199 at commit [`a069d01`](https://github.com/apache/spark/commit/a069d01a485b6066650b75e0cb9bfaf63710a9e2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93801/
Test FAILed.

HeartSaVioR commented:
@arunmahadevan Thanks for rebasing. I'll take a look.

HeartSaVioR reviewed (commented):
The code change looks good overall. 

Left comments mostly suggesting about deduplicating code between micro-batch and continuous and kind of defensive programming. Some comments could be shown as individual's preference and I'm definitely sure to follow Spark's preference.

arunmahadevan committed:
SPARK-24127: address review comments

arunmahadevan commented:
@HeartSaVioR , Addressed your comments. Let me know if I missed something. Also rebased and had to change more code to use the new interfaces. 

I hope if we can speed up the review cycles in general than leaving PRs to hibernation for a while and then the developer will loose the context and other things would have changed in the meanwhile.

SparkQA commented:
**[Test build #93921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93921/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

jose-torres commented:
The change looks broadly good (and important) to me. I'll defer to @HeartSaVioR wrt the in-depth review; let me know if there are any specific parts I should to take a look at.

SparkQA commented:
**[Test build #93921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93921/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93921/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #93939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93939/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala:
```diff
@@ -0,0 +1,295 @@
+      .sortBy(_.partitionId)
+      .map(_.offset)
+      .toList
+    TextSocketOffset(offs)
```
Yeah, agreed. I'm OK if same implications are used in other places.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  private def initialize(): Unit = synchronized {
+    socket = new Socket(host, port)
```
Yeah you're planning to investigate and touch APIs then it sounds really good. May worth to file a new issue?

arunmahadevan replied:
filed: https://issues.apache.org/jira/browse/SPARK-25000

HeartSaVioR reviewed (approved):
LGTM given we are planning to tackle deduplication of codebase between micro-batch and continuous later.

SparkQA commented:
**[Test build #93939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93939/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93939/
Test FAILed.

arunmahadevan commented:
retest this please

SparkQA commented:
**[Test build #94011 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94011/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

SparkQA commented:
**[Test build #94011 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94011/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94011/
Test FAILed.

arunmahadevan commented:
retest this please

SparkQA commented:
**[Test build #94148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94148/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

SparkQA commented:
**[Test build #94148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94148/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94148/
Test FAILed.

arunmahadevan commented:
retest this please

SparkQA commented:
**[Test build #94321 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94321/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

SparkQA commented:
**[Test build #94321 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94321/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94321/
Test PASSed.

arunmahadevan commented:
@HyukjinKwon this has been open for a while, would you mind taking this forward?

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #94411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94411/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).

HyukjinKwon commented:
@jose-torres and @HeartSaVioR, is it good to go?

SparkQA commented:
**[Test build #94411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94411/testReport)** for PR 21199 at commit [`f4a39d9`](https://github.com/apache/spark/commit/f4a39d9ebae2d6f6ae59caf3140310b17e75b602).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TextSocketContinuousReader(options: DataSourceOptions) extends ContinuousReader with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94411/
Test PASSed.

HyukjinKwon commented:
Merged to master.
